rfcs/0116-artifact-service.md

```diff
@@ -0,0 +1,196 @@
+# RFC 116 Artifact Service
+* Comments: [#116](https://api.github.com/repos/taskcluster/taskcluster-rfcs/issues/116)
+* Proposed by: @jhford
+
+# Summary
+
+Taskcluster tasks often product Artifacts during the course of being run.
+Examples include browser archives, json metadata and log files.  We currently
+store Artifacts in a single region of S3 and split management of Artifacts
+between the Queue and Cloud Mirror.  This project is to take Artifact handling
+out of the Queue and into its own service that takes over the responsibilities
+for cross-region and cross-cloud replication.
+
+The Artifact service should not be directly exposed to end users as a Rest API.
+It will instead be a generalised API that the Queue uses to facilitate artifact
+creation from tasks.  In other words, the Queue gates operations to the
+Artifact service.  The service will be exposed to endusers by means of a `Location`
+header value on a `300` series redirect.
+
+The Queue will now store only the name of Artifacts and would treat the name of
+an Artifact as an opaque identifier and defer all management of the actual
+bytes to this system.  The Queue would also stop doing IP to Object store
+mappings.
+
+This RFC is to decide on what the Artifacts service is and its interface, not
+implementation details
+
+## Motivation
+
+* We want the Queue to be easier to understand, develop and maintain
+* We want Artifacts to be easier to understand, develop and maintain
+* We want better Artifact handling without increasing complexity of the Queue
+  codebase
+* We want to have more clear responsibilities for our services
+* We want to consolidate the handling of Artifacts into a single service
+* We want to build generalised software when possible as participants of a free
+  software community
+
+# Details
+
+* The Artifact service will start from the Queue's `artifacts.js` file
+* Auth might not be done with scopes and tc-auth, but a shared secret between
+  Queue and Artifacts
```

Comment:
We need to settle on a specific auth mechanism.  JWT looks fine to me so far, I'll play around with it some.

We should block overwriting of artifacts in general, since that's not likely something we actually want ever.

I presume the namespace of this service here to not really be a 'task' based namespace, but that the Queue would internally implement its own 'task/run/artifactName' namespacing.  If someone got the secret from this service, it would almost certainly be stored beside the same secrets that would allow for raw access to the storage system and so isn't really a concern.

That secret could also be leaked from the Queue, which wouldn't have the underlying storage credentials.  That would then allow the bearer of that token to create new artifacts with any name.  That would allow someone to pre-emptively create an artifact for a task, but the real task would fail if we block overwriting and would alert us to the name being already taken.

At some level though, a secret which allows an operation must be kept secret.  This is no more risky than S3 itself.  Right now, the secret used in the Queue is a set of raw S3 credentials.  If you lose your access token and client id, someone can change your stored objects.